src/tests/sourcepos.rs

```diff
@@ -0,0 +1,512 @@
+[bar]: https://example.com"#,
+);
+
+const IMAGE: TestCase = (
+    &[sourcepos!((1:7-1:38))],
+    "hello ![alt text](https://example.com) banana",
+);
+
+const MATH: TestCase = (
+    &[
+        sourcepos!((3:1-3:7)),
+        sourcepos!((3:17-3:26)),
+        sourcepos!((3:36-3:44)),
+    ],
+    r#"hello
+
+$1 + 1$ between $`1 + 23`$ between $$a + b$$
+
+banana"#,
+);
+
+const ESCAPED: TestCase = (
+    &[
+        sourcepos!((1:1-1:2)),
+        sourcepos!((1:3-1:4)),
+        sourcepos!((1:5-1:6)),
+        sourcepos!((1:7-1:8)),
+        sourcepos!((1:9-1:10)),
+        sourcepos!((1:11-1:12)),
+        sourcepos!((1:13-1:14)),
+        sourcepos!((1:15-1:16)),
+        sourcepos!((1:17-1:18)),
+        sourcepos!((1:19-1:20)),
+        sourcepos!((1:21-1:22)),
+        sourcepos!((1:23-1:24)),
+        sourcepos!((1:25-1:26)),
+        sourcepos!((1:27-1:28)),
+        sourcepos!((1:29-1:30)),
+        sourcepos!((1:31-1:32)),
+        sourcepos!((1:33-1:34)),
+        sourcepos!((1:35-1:36)),
+        sourcepos!((1:37-1:38)),
+        sourcepos!((1:39-1:40)),
+        sourcepos!((1:41-1:42)),
+        sourcepos!((1:43-1:44)),
+        sourcepos!((1:45-1:46)),
+        sourcepos!((1:47-1:48)),
+        sourcepos!((1:49-1:50)),
+        sourcepos!((1:51-1:52)),
+        sourcepos!((1:53-1:54)),
+        sourcepos!((1:55-1:56)),
+        sourcepos!((1:57-1:58)),
+        sourcepos!((1:59-1:60)),
+        sourcepos!((1:61-1:62)),
+        sourcepos!((1:63-1:64)),
+    ],
+    r#"\!\"\#\$\%\&\'\(\)\*\+\,\-\.\/\:\;\<\=\>\?\@\[\\\]\^\_\`\{\|\}\~\a"#,
+);
+
+const WIKI_LINK: TestCase = (
+    &[sourcepos!((1:1-1:9)), sourcepos!((3:1-3:33))],
+    r#"[[floop]]
+
+[[http://example.com|some title]]
+
+after"#,
+);
+
+const UNDERLINE: TestCase = (&[sourcepos!((1:8-1:22))], "before __hello world__ after");
+
+const SPOILERED_TEXT: TestCase = (
+    &[sourcepos!((2:1-2:11))],
+    r#"before
+||spoiler||
+after"#,
+);
+
+const ESCAPED_TAG: TestCase = (
+    &[sourcepos!((2:1-2:8))],
+    r#"before
+||hello|
+after"#,
+);
+
+fn node_values() -> HashMap<NodeValueDiscriminants, TestCase> {
+    use NodeValueDiscriminants::*;
+
+    NodeValueDiscriminants::VARIANTS
+        .iter()
+        .filter(|v| {
+            !matches!(
+                v,
+                // Remove buggy variants.
+                List // end is 3:0
+                    | Item // end is 3:0
+                    | TaskItem // end is 4:0
+                    | DescriptionItem // end is 4:0
+                    | DescriptionTerm // end is 3:0
+                    | DescriptionDetails // end is 4:0
+                    | HtmlInline // end is 1:31 but should be 3:14
+                    | LineBreak // start is 1:15 but should be 1:13
+                    | Code // is 1:8-1:12 but should be 1:7-1:13
+                    | ThematicBreak // end is 4:0
+                    | Link // inconsistent between link types
+                    | Math // is 3:2-3:6 but should be 3:1-3:7
```

Comment:
If y'all can confirm that all of these are&mdash;in fact&mdash;bugs, I'll try my hand at fixing them.

Reply:
Yeah I think those are all incorrect sourcesos 😢